core/src/main/java/io/grpc/internal/ServerCallImpl.java

```diff
@@ -221,6 +222,17 @@ public ServerStreamListenerImpl(
       this.call = checkNotNull(call, "call");
       this.listener = checkNotNull(listener, "listener must not be null");
       this.context = checkNotNull(context, "context");
+      // Wire ourselves up so that if the context is cancelled, our flag call.cancelled also
+      // reflects the new state. Use a DirectExecutor so that it happens in the same thread
+      // as the caller of {@link Context#cancel}.
+      this.context.addListener(
+          new Context.CancellationListener() {
+            @Override
+            public void cancelled(Context context) {
+              ServerStreamListenerImpl.this.call.cancelled = true;
```

Comment:
If this becomes a problem, we could move it to ServerStream by having ServerCall.isCancelled call ServerStream.isCancelled (a new method).